File: src/jdwp/idCodec.ts
// Big-endian byte reader/writer for the JDWP wire protocol. JDWP sizes its
// reference IDs (objectID, methodID, referenceTypeID, fieldID, frameID) per the
// VM's VirtualMachine.IDSizes reply, so reads/writes of those take the
// negotiated width; every other field is fixed-width big-endian. IDs are kept
// as bigint to hold the full 8-byte range losslessly.
//
// Port reference for togo/internal/jdwp/idcodec.go.

export interface IdSizes {
  fieldID: number;
  methodID: number;
  objectID: number;
  referenceTypeID: number;
  frameID: number;
}

// Before VirtualMachine.IDSizes is read only fixed-width fields are decodable;
// 8 bytes is the modern HotSpot default and a safe placeholder until the real
// sizes arrive.
export const DEFAULT_ID_SIZES: IdSizes = {
  fieldID: 8,
  methodID: 8,
  objectID: 8,
  referenceTypeID: 8,
  frameID: 8,
};

export class ByteWriter {
  private readonly chunks: Buffer[] = [];

  u1(n: number): this {
    this.chunks.push(Buffer.from([n & 0xff]));
    return this;
  }
  u2(n: number): this {
    const b = Buffer.allocUnsafe(2);
    b.writeUInt16BE(n & 0xffff, 0);
    this.chunks.push(b);
    return this;
  }
  u4(n: number): this {
    const b = Buffer.allocUnsafe(4);
    b.writeUInt32BE(n >>> 0, 0);
    this.chunks.push(b);
    return this;
  }
  i4(n: number): this {
    const b = Buffer.allocUnsafe(4);
    b.writeInt32BE(n | 0, 0);
    this.chunks.push(b);
    return this;
  }
  u8(n: bigint): this {
    return this.id(n, 8);
  }
  boolean(b: boolean): this {
    return this.u1(b ? 1 : 0);
  }
  /** A `size`-byte big-endian ID (size comes from the negotiated IdSizes). */
  id(value: bigint, size: number): this {
    const b = Buffer.allocUnsafe(size);
    let v = value;
    for (let i = size - 1; i >= 0; i--) {
      b[i] = Number(v & 0xffn);
      v >>= 8n;
    }
    this.chunks.push(b);
    return this;
  }
  /** JDWP string: u4 length of the UTF-8 bytes, then the bytes (no NUL). */
  string(s: string): this {
    const bytes = Buffer.from(s, "utf8");
    this.u4(bytes.length);
    this.chunks.push(bytes);
    return this;
  }
  bytes(buf: Buffer): this {
    this.chunks.push(buf);
    return this;
  }
  get length(): number {
    return this.chunks.reduce((n, c) => n + c.length, 0);
  }
  toBuffer(): Buffer {
    return Buffer.concat(this.chunks);
  }
}

export class ByteReader {
  private offset = 0;
  constructor(private readonly buf: Buffer) {}

  u1(): number {
    return this.buf.readUInt8(this.offset++);
  }
  u2(): number {
    const v = this.buf.readUInt16BE(this.offset);
    this.offset += 2;
    return v;
  }
  u4(): number {
    const v = this.buf.readUInt32BE(this.offset);
    this.offset += 4;
    return v;
  }
  i4(): number {
    const v = this.buf.readInt32BE(this.offset);
    this.offset += 4;
    return v;
  }
  u8(): bigint {
    return this.id(8);
  }
  boolean(): boolean {
    return this.u1() !== 0;
  }
  /** Read a `size`-byte big-endian ID as a bigint. */
  id(size: number): bigint {
    let v = 0n;
    for (let i = 0; i < size; i++) v = (v << 8n) | BigInt(this.buf[this.offset++]);
    return v;
  }
  string(): string {
    const len = this.u4();
    const s = this.buf.toString("utf8", this.offset, this.offset + len);
    this.offset += len;
    return s;
  }
  bytes(n: number): Buffer {
    const b = this.buf.subarray(this.offset, this.offset + n);
    this.offset += n;
    return b;
  }
  get remaining(): number {
    return this.buf.length - this.offset;
  }
}

File: togo/internal/jdwp/idcodec.go
// Package jdwp implements a JDWP (Java Debug Wire Protocol) client for cappu's
// debug adapter: byte/packet codecs, the connection, typed command wrappers and
// event decoding.
package jdwp

// Big-endian byte reader/writer for the JDWP wire protocol. JDWP sizes its
// reference IDs (objectID, methodID, ...) per the VM's VirtualMachine.IDSizes
// reply, so reads/writes of those take the negotiated width; every other field
// is fixed-width big-endian. IDs are kept as uint64 to hold the full 8-byte
// range. Port of src/jdwp/idCodec.ts.

import "encoding/binary"

// IDSizes is the per-VM byte width of each JDWP reference id kind.
type IDSizes struct {
	FieldID         int
	MethodID        int
	ObjectID        int
	ReferenceTypeID int
	FrameID         int
}

// DefaultIDSizes is the placeholder used before VirtualMachine.IDSizes is read
// (8 bytes is the modern HotSpot default).
var DefaultIDSizes = IDSizes{FieldID: 8, MethodID: 8, ObjectID: 8, ReferenceTypeID: 8, FrameID: 8}

// Writer accumulates big-endian JDWP fields.
type Writer struct{ buf []byte }

func (w *Writer) U1(n byte) *Writer { w.buf = append(w.buf, n); return w }

func (w *Writer) U2(n uint16) *Writer {
	w.buf = binary.BigEndian.AppendUint16(w.buf, n)
	return w
}

func (w *Writer) U4(n uint32) *Writer {
	w.buf = binary.BigEndian.AppendUint32(w.buf, n)
	return w
}

func (w *Writer) I4(n int32) *Writer { return w.U4(uint32(n)) }

func (w *Writer) U8(n uint64) *Writer { return w.ID(n, 8) }

func (w *Writer) Bool(b bool) *Writer {
	if b {
		return w.U1(1)
	}
	return w.U1(0)
}

// ID writes a size-byte big-endian id (size from the negotiated IDSizes).
func (w *Writer) ID(v uint64, size int) *Writer {
	tmp := make([]byte, size)
	for i := size - 1; i >= 0; i-- {
		tmp[i] = byte(v & 0xff)
		v >>= 8
	}
	w.buf = append(w.buf, tmp...)
	return w
}

// String writes a JDWP string: u4 byte length of the UTF-8 bytes, then the bytes.
func (w *Writer) String(s string) *Writer {
	w.U4(uint32(len(s)))
	w.buf = append(w.buf, s...)
	return w
}

func (w *Writer) Bytes(b []byte) *Writer { w.buf = append(w.buf, b...); return w }

func (w *Writer) Len() int       { return len(w.buf) }
func (w *Writer) Buffer() []byte { return w.buf }

// Reader consumes big-endian JDWP fields from a packet body.
type Reader struct {
	buf []byte
	off int
}

func NewReader(b []byte) *Reader { return &Reader{buf: b} }

func (r *Reader) U1() byte {
	v := r.buf[r.off]
	r.off++
	return v
}

func (r *Reader) U2() uint16 {
	v := binary.BigEndian.Uint16(r.buf[r.off:])
	r.off += 2
	return v
}

func (r *Reader) U4() uint32 {
	v := binary.BigEndian.Uint32(r.buf[r.off:])
	r.off += 4
	return v
}

func (r *Reader) I4() int32 { return int32(r.U4()) }

func (r *Reader) U8() uint64 { return r.ID(8) }

func (r *Reader) Bool() bool { return r.U1() != 0 }

// ID reads a size-byte big-endian id.
func (r *Reader) ID(size int) uint64 {
	var v uint64
	for range size {
		v = v<<8 | uint64(r.buf[r.off])
		r.off++
	}
	return v
}

func (r *Reader) String() string {
	n := int(r.U4())
	s := string(r.buf[r.off : r.off+n])
	r.off += n
	return s
}

func (r *Reader) Bytes(n int) []byte {
	b := r.buf[r.off : r.off+n]
	r.off += n
	return b
}

func (r *Reader) Remaining() int { return len(r.buf) - r.off }
